feat: transform imagePullPolicy when using local cluster

lucasrod16 opened the pull request:
<!-- Thank you for your contribution! -->

<!-- Include if applicable: -->
Relates to #6992

**Description**
<!-- Describe your changes here. The more detail, the easier the review! -->
Introduces a `ReplaceImagePullPolicy` function to update the `imagePullPolicy` field in Kubernetes manifests to `Never` when running on a local cluster. This prevents deployment errors that occur when `imagePullPolicy` is set to `Always`, as detailed in #6992.

**User facing changes**

Files to reproduce:

- Dockerfile

  ```Dockerfile
  FROM nginx:alpine
  ```
- pod.yaml

  ```yaml
  apiVersion: v1
  kind: Pod
  metadata:
    name: nginx
  spec:
    containers:
    - name: nginx
      image: nginx:alpine
      imagePullPolicy: Always
  ```
- skaffold.yaml

  ```yaml
  apiVersion: skaffold/v4beta11
  kind: Config
  metadata:
    name: nginx
  build:
    artifacts:
      - image: nginx
        context: .
        docker:
          dockerfile: Dockerfile
  manifests:
    rawYaml:
     - ./pod.yaml
  ```

Before:
<img width="1302" alt="failed" src="https://github.com/user-attachments/assets/dc1f82e3-4986-4646-892a-5a286e5aa88d">

After:
<img width="1010" alt="success" src="https://github.com/user-attachments/assets/cda4314a-845f-4bc1-87ca-e6609a1a0f1a">

<!--
Please be sure your PR includes unit tests - we don't merge code that brings down test coverage! 
Integration tests are sometimes an appropriate substitute. 
-->

lucasrod16 commented:
Hi @ericzzzzzzz,

I noticed that this pull request has been pending for a while. Is there anything needed from me? Please let me know if there are any updates required.

Thanks!

ericzzzzzzz commented:
Hi, @lucasrod16  Thank you for the pr! 

I think we should only do this transformation when using `skaffold dev`, `skaffold run`, `skaffold debug` as we know that the transformed manifests will be applied to the connected local cluster, but for `skaffold render` command, we're not sure how the manifests will be used, the rendered results may be used in other places even without `skaffold`. 


Also, this pr does the requested change when kubectl deployer is used, it needs to do the similar thing when helm deployer is used.

lucasrod16 commented:
> Hi, @lucasrod16 Thank you for the pr!
> 
> I think we should only do this transformation when using `skaffold dev`, `skaffold run`, `skaffold debug` as we know that the transformed manifests will be applied to the connected local cluster, but for `skaffold render` command, we're not sure how the manifests will be used, the rendered results may be used in other places even without `skaffold`.
> 
> Also, this pr does the requested change when kubectl deployer is used, it needs to do the similar thing when helm deployer is used.

@ericzzzzzzz That makes sense, thanks for the feedback!

I moved the transform logic from `BaseTransform` to kubectl `Deploy` and verified it works locally with `dev`, `run`, and `debug`.

I will work on adding the transform logic to the helm deployer as well

alphanota commented on pkg/skaffold/kubernetes/manifest/image_pull_policy.go:
```diff
@@ -0,0 +1,79 @@
+
+import apimachinery "k8s.io/apimachinery/pkg/runtime/schema"
+
+type ResourceSelectorImagePullPolicy struct{}
```
This should be unexported, since this new struct is only used as an implementor of the ResourceSelector interface. 

I would also add a comment about what this ResourceSelector does/ selects. For example
resourceSelectorImagePullPolicy is an implementation of ResourceSelector to select X and do Z.

alphanota commented on pkg/skaffold/kubernetes/manifest/image_pull_policy.go:
```diff
@@ -0,0 +1,79 @@
+
+type ResourceSelectorImagePullPolicy struct{}
+
+func NewResourceSelectorImagePullPolicy() *ResourceSelectorImagePullPolicy {
```
Should return the interface ResourceSelector instead of *ResourceSelectorImagePullPolicy.

alphanota commented on pkg/skaffold/kubernetes/manifest/image_pull_policy.go:
```diff
@@ -0,0 +1,79 @@
+	return &ResourceSelectorImagePullPolicy{}
+}
+
+func (rs *ResourceSelectorImagePullPolicy) allowByGroupKind(gk apimachinery.GroupKind) bool {
```
A short comment explaining the goal of the method and the criteria used would be helpful.

alphanota commented on pkg/skaffold/kubernetes/manifest/image_pull_policy.go:
```diff
@@ -0,0 +1,79 @@
+	return false
+}
+
+func (rs *ResourceSelectorImagePullPolicy) allowByNavpath(gk apimachinery.GroupKind, navpath string, k string) (string, bool) {
```
Here too, a description of the goal would help.

alphanota commented on pkg/skaffold/kubernetes/manifest/image_pull_policy.go:
```diff
@@ -0,0 +1,79 @@
+	return l.Visit(r, rs)
+}
+
+type imagePullPolicyReplacer struct{}
```
A comment explaining the purpose of this struct. ie imagePullPolicyReplacer implements FieldVisitor and is used to edit an entry in a YAML document.

alphanota commented on pkg/skaffold/kubernetes/manifest/image_pull_policy.go:
```diff
@@ -0,0 +1,79 @@
+
+type imagePullPolicyReplacer struct{}
+
+func (i *imagePullPolicyReplacer) Visit(gk apimachinery.GroupKind, navpath string, o map[string]interface{}, k string, v interface{}, rs ResourceSelector) bool {
```
Please add a comment explaining this method also. For example Visit sets the value of the "imagePullPolicy" field in a Kubernetes manifest to "Never".

alphanota reviewed (changes requested):
Hello! Thank  you for sending this PR! Please see my comments on it, thanks!.

Also for the PR description, since this PR doesn't have the helm changes needed to fully address the issue, I don't think we should close the bug when this PR is merged.

mepi262 commented:
@lucasrod16 
The reviewer has commented to your pull request. How about check it?

lucasrod16 commented:
Hey @alphanota, thanks for the PR review! I've updated the PR description to ensure the related issue isn't closed when this PR is merged and have addressed all your feedback. I currently don't have the bandwidth to implement the helm functionality, and I appreciate your understanding. Thanks again for the feedback!

plumpy commented:
Given the issue reported in #9687, this appears to cause a breaking change. I'm going to go ahead and revert this PR.

I _suspect_ what needs to happen is you only want to replace the pull policy on images that were actually built by Skaffold as part of a `skaffold dev` command. But it might need more discussion.
